### PR TITLE
Deprecated notice fired by wp_set_object_terms() with PHP 8.1

### PR DESCRIPTION
**Pull request by peterwilsoncc**
Prevents the throwing of a notice in `wp_set_object_terms()`.

`src` changes props @SergeyBiryukov 

Trac ticket: https://core.trac.wordpress.org/ticket/57923

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/term/wpSetObjectTerms.php [hunk 1]
```diff
@@ -430,4 +432,47 @@ public function test_non_existent_integers_should_be_ignored() {
 
 		$this->assertSame( array(), $tt_ids );
 	}
+
+	/**
+	 * Empty values should clear an object of all terms.
```
Comment (costdev) [inline tests/phpunit/tests/term/wpSetObjectTerms.php]:
```suggestion
	 * Tests that empty values clear an object of all terms.
```

File: tests/phpunit/tests/term/wpSetObjectTerms.php [hunk 2]
```diff
@@ -430,4 +432,47 @@ public function test_non_existent_integers_should_be_ignored() {
...
+	 *
+	 * @ticket 57923
+	 *
+	 * @dataProvider data_empty_value_should_clear_terms
+	 *
+	 * @param mixed $empty_value An empty.
```
Comment (costdev) [inline tests/phpunit/tests/term/wpSetObjectTerms.php]:
```suggestion
	 * @param mixed $empty_value An empty value.
```

File: tests/phpunit/tests/term/wpSetObjectTerms.php [hunk 3]
```diff
@@ -430,4 +432,47 @@ public function test_non_existent_integers_should_be_ignored() {
...
+	 */
+	public function test_empty_value_should_clear_terms( $empty_value ) {
+		$post_id = self::$post_ids[0];
+
+		// Assign some terms.
+		wp_set_object_terms( $post_id, self::$term_ids, self::$taxonomy );
+
+		// Make sure the terms are set.
+		$terms = wp_get_object_terms( $post_id, self::$taxonomy, array( 'fields' => 'names' ) );
+		$this->assertNotEmpty( $terms );
```
Comment (costdev) [inline tests/phpunit/tests/term/wpSetObjectTerms.php]: These assertions should have a `$message` parameter.

File: tests/phpunit/tests/term/wpSetObjectTerms.php [hunk 4]
```diff
@@ -430,4 +432,47 @@ public function test_non_existent_integers_should_be_ignored() {
...
+		// Remove terms by passing an empty value.
+		wp_set_object_terms( $post_id, $empty_value, self::$taxonomy );
+
+		// Make sure the terms have been removed.
+		$terms = wp_get_object_terms( $post_id, self::$taxonomy, array( 'fields' => 'names' ) );
+		$this->assertEmpty( $terms );
+	}
+
+	/**
+	 * Data provider for test_empty_value_should_clear_terms().
```
Comment (costdev) [inline tests/phpunit/tests/term/wpSetObjectTerms.php]:
```suggestion
	 * Data provider.
```

We're currently using this for data provider docblocks until consensus is reached on a consistent format.

```php
/**
 * Data provider.
 *
 * @return array[]
 */
```

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the PR @peterwilsoncc!

I've added some feedback in this review.

File: tests/phpunit/tests/term/wpSetObjectTerms.php (hunk 3 above)
Comment (peterwilsoncc) [inline tests/phpunit/tests/term/wpSetObjectTerms.php]:
```suggestion
		$this->assertNotEmpty( $terms, 'Terms should initially be applied to post object.' );
```

File: tests/phpunit/tests/term/wpSetObjectTerms.php [hunk 5]
```diff
@@ -430,4 +432,47 @@ public function test_non_existent_integers_should_be_ignored() {
...
+		// Make sure the terms are set.
+		$terms = wp_get_object_terms( $post_id, self::$taxonomy, array( 'fields' => 'names' ) );
+		$this->assertNotEmpty( $terms );
+
+		// Remove terms by passing an empty value.
+		wp_set_object_terms( $post_id, $empty_value, self::$taxonomy );
+
+		// Make sure the terms have been removed.
+		$terms = wp_get_object_terms( $post_id, self::$taxonomy, array( 'fields' => 'names' ) );
+		$this->assertEmpty( $terms );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/term/wpSetObjectTerms.php]:
```suggestion
		$this->assertEmpty( $terms, 'An empty() value should clear terms from the post object.' );
```

**Comment by peterwilsoncc** [comment]
Thanks @costdev , I've applied each of your suggestions.

**Review by costdev — APPROVED** [review]
Thanks for the updates @peterwilsoncc! LGTM 👍

**Comment by peterwilsoncc** [comment]
Merged in [r55921](https://core.trac.wordpress.org/changeset/55921) / https://github.com/WordPress/wordpress-develop/commit/6178829ab0594d3a541c406502ba31e05ddd82f3

**Pull request closed without merging**